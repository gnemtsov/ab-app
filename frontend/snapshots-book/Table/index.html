
            <!doctype html>
            <html lang="en">
                <head>
                    <meta charset="utf-8">
                    <title>Table</title>
                    <style>
            .TOCContainer a{
                display:block;
                font-size: 1.2rem; 
                margin-left: 1rem;
            }

            .SubHeader{ padding: 0.6rem; background-color: #d1d1d1; }
            .Filters span { margin-left: 1rem; }

            .TestResultContainer{ margin-top: 1.5rem; }
            .TestResultContainer.passed .TestHeader{ padding: 0.6rem; background-color: #99e99b; }
            .TestResultContainer.failed .TestHeader{ padding: 0.6rem; background-color: #ff9f9f; }
            .TestResult{
                display: flex;
                justify-content: space-around;                
            }

            .ExpectedContainer, .ActualContainer {
                padding: 0.6rem;
                width: 100%;
            }
            .ExpectedHeader, .ActualHeader { color: #4c4c4c; }
            .ExpectedContainer iframe, .ActualContainer iframe{
                width: 100%;
                height: 350px;
                border: 1px solid #beb6b6;
            }
        </style>
                    <script>
            document.addEventListener("DOMContentLoaded", function(){
                document.getElementById('show-all').addEventListener("click", function(e) {
                    e.preventDefault();
                    var resultContainers = document.getElementsByClassName("TestResultContainer");
                    for (var i = 0; i < resultContainers.length; ++i) {
                        var el = resultContainers[i];
                        el.style.display = 'block';
                    }                    
                });

                document.getElementById('show-passed').addEventListener("click", function(e) {
                    e.preventDefault();
                    var resultContainers = document.getElementsByClassName("TestResultContainer");
                    for (var i = 0; i < resultContainers.length; ++i) {
                        var el = resultContainers[i];
                        if(el.classList.contains('passed')){
                            el.style.display = 'block';
                        } else {
                            el.style.display = 'none';
                        }
                    }                    
                });

                document.getElementById('show-failed').addEventListener("click", function(e) {
                    e.preventDefault();
                    var resultContainers = document.getElementsByClassName("TestResultContainer");
                    for (var i = 0; i < resultContainers.length; ++i) {
                        var el = resultContainers[i];
                        if(el.classList.contains('failed')){
                            el.style.display = 'block';
                        } else {
                            el.style.display = 'none';
                        }
                    };
                });               
            });
        </script>
                </head>
                <body>
                <div><a href="../index.html">Table of contents</a></div>
                <h1>Table</h1>
                <div class="Filters">
                    Show 
                    <span><a id="show-all" href="#">all</a>(1)</span>
                    <span><a id="show-passed" href="#">passed</a>(1)</span>
                    <span><a id="show-failed" href="#">failed</a>(0)</span>
                </div>
                
                    <div class="TestResultContainer passed">
                        <div class="TestHeader">Basic render test</div>
                        <div class="TestResult">
                            
                    <div class="ExpectedContainer">
                        <div class="ExpectedHeader"></div>
                        <iframe src="1/expectedHtml.html">
                            Browser should support iframes.
                        </iframe>
                    </div>
                
                            
                        </div>
                    </div>
                
            </body>
            </html>
        