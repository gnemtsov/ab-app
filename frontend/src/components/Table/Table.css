.Container {
    display: grid;
    grid-template-columns: max-content min-content;
}

.Table {
    border-spacing: 0;
}

.Table thead th {
    text-align: left;
}

.Table tbody td,
.Table thead th {
    padding: 0.4rem;
}

.Odd {
    background-color: #e9e8e8;
}

.Selected {
    background-color: aquamarine;
}

.Footer {
    display: flex;
    justify-content: space-between;
    align-items: center;
    flex-wrap: wrap-reverse;
}

.Footer div {
    margin-top: 1rem;
}

.Footer div.Legend {
    font-size: 0.9rem;
}

.Footer div:first-child {
    margin-right: 2rem;
}