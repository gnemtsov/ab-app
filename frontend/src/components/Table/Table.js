import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Paginator from './Paginator/Paginator';
import Toolbar from './Toolbar/Toolbar';
import classes from './Table.css';

export default class Table extends Component {
    defaultSortParams = []

    static defaultProps = {
        rowsPerPage: 10,
        selectable: false,
        csvExport: true,
        emptyTableMessage: 'No data specified',
        cols: [],
        rows: []
    }

    static propTypes = {
        rowsPerPage: PropTypes.number,
        selectable: PropTypes.bool,
        csvExport: PropTypes.bool,
        emptyTableMessage: PropTypes.string,
        cols: PropTypes.array,
        rows: PropTypes.array
    }

    state = {
        cols: [],
        rows: [],
        sortParams: [],
        selected: [],
        totalPages: 1,
        currentPage: 1,
        toolbarShow: false,
        isEmpty: true,
        bodyTop: 0,
        headerHeight: 0,
        bodyHeight: 0
    }

    constructor(props) { //constructor updates initial state
        super(props);

        this.refHeader = React.createRef()
        this.refBody = React.createRef()

        let { rowsPerPage, cols, rows } = props;
        const totalRows = rows.length;
        const totalPages = Math.ceil(totalRows / rowsPerPage);

        //make defaultSortParams
        let sortCols = cols.filter(col => col.sortDirection !== undefined);
        if (sortCols.length) {
            sortCols.sort((a, b) => a.sortOrder - b.sortOrder);

            this.defaultSortParams = sortCols.map((col) => ({
                name: col.name,
                dir: col.sortDirection
            }));
        }

        rows = this.multiSort(rows, this.defaultSortParams); //sort rows

        this.state = {
            ...this.state,
            cols,
            rows,
            isEmpty: !cols.length || !rows.length,
            sortParams: this.defaultSortParams,
            totalPages
        };
    }

    componentDidMount() {
        if (!this.state.isEmpty) {
            const bodyRect = this.refBody.current.getBoundingClientRect();
            const headerRect = this.refHeader.current.getBoundingClientRect();

            this.setState({
                bodyTop: bodyRect.top + window.scrollY,
                headerHeight: headerRect.height,
                bodyHeight: bodyRect.height
            });
        }
    }


    multiSort = (arr, sortParams) => {
        if (!sortParams.length) {
            return arr.slice();
        }

        const cols = sortParams.map((col) => col.name);
        const dirs = sortParams.map((col) => col.dir);

        const sortRecursive = (a, b, cols, dirs, index) => {
            const col = cols[index];
            const dir = dirs[index];
            let x = a.data[col];
            let y = b.data[col];

            if (typeof x === 'string' || typeof y === 'string') {
                x = x === null ? '' : x.toLowerCase();
                y = y === null ? '' : y.toLowerCase();
            }

            if (x < y) {
                return dir === 'DESC' ? 1 : -1;
            }

            if (x > y) {
                return dir === 'DESC' ? -1 : 1;
            }

            return cols.length - 1 > index ? sortRecursive(a, b, cols, dirs, index + 1) : a.i - b.i;
        }

        let sortArr = arr.map((data, i) => ({ data, i })); //mapping needed to make sort stable for equal values
        sortArr.sort((a, b) => sortRecursive(a, b, cols, dirs, 0));
        return sortArr.map(el => el.data);
    }

    headerMouseDownHandler = (event, colName) => { //sort handler
        event.preventDefault();

        let sortParams = this.state.sortParams.slice();
        let sortIndex = sortParams.findIndex(el => el.name === colName);
        if (event.shiftKey) {
            if (sortIndex === -1) {
                sortParams.push({
                    name: colName,
                    dir: 'ASC'
                });
            } else if (sortParams[sortIndex].dir === 'ASC') {
                sortParams[sortIndex].dir = 'DESC';
            } else {
                sortParams.splice(sortIndex, 1);
            }
        } else {
            if (sortIndex === -1) {
                sortParams = [{
                    name: colName,
                    dir: 'ASC'
                }];
            } else if (sortParams[sortIndex].dir === 'ASC') {
                sortParams = [{
                    name: colName,
                    dir: 'DESC'
                }];
            } else {
                sortParams = [];
            }
        }

        const sortedRows = this.multiSort(this.state.rows, sortParams); //sort rows
        const selected = this.state.selected.map(i => sortedRows.indexOf(sortedRows[i])); //remap selected values

        this.setState({
            rows: sortedRows,
            selected: selected,
            sortParams: sortParams
        });
    }

    rowMouseDownHandler = (event, row, pageFirstRow, pageLastRow) => { //select handler
        let selected = this.state.selected.slice();

        if (event.ctrlKey) {
            const selectedRowIndex = selected.indexOf(row);
            selectedRowIndex === -1 ? selected.push(row) : selected.splice(selectedRowIndex, 1);
        } else if (event.shiftKey) {
            event.preventDefault();
            const selectedRowIndex = selected.indexOf(row);
            const selectedOnPage = selected.filter(i => i >= pageFirstRow && i <= pageLastRow);
            const lastSelectedRow = selectedOnPage.length ? selectedOnPage[selectedOnPage.length - 1] : row;

            const selectAction = (i) => {
                const selectedI = selected.indexOf(i);
                if (selectedRowIndex === -1) {
                    selectedI === -1 && selected.push(i);
                } else {
                    row !== i && selected.splice(selectedI, 1);
                }
            }

            if (row > lastSelectedRow) {
                for (let i = lastSelectedRow; i <= row; i++) {
                    selectAction(i);
                }
            } else {
                for (let i = lastSelectedRow; i >= row; i--) {
                    selectAction(i);
                }
            }
        } else {
            selected = selected.filter(row => row < pageFirstRow || row > pageLastRow);
            selected.push(row);
        }

        this.setState({ selected: selected });
    }

    pageClickHandler = (event, button) => { //paginator handler
        event.preventDefault();
        this.setState(prevState => {
            switch (button) {
                case 'back': return { currentPage: prevState.currentPage - 1 };
                case 'forward': return { currentPage: prevState.currentPage + 1 };
                default: return { currentPage: button }
            }
        })
    }

    toolbarShow = () => this.setState({ toolbarShow: true })
    toolbarHide = () => this.setState({ toolbarShow: false })

    csvExportHandler = (event) => {
        const { cols, rows } = this.state;

        let table = 'No data specified';
        if (cols.length) {
            const totalRows = rows.length;

            //head
            const thead = cols.map(col => '"' + String(col.title) + '"').join(';');

            //body
            let tbody = [];
            for (let i = 0; i < totalRows; i++) {
                let cells = [];
                const row = rows[i];
                for (var col in row) {
                    if (row.hasOwnProperty(col)) {
                        cells.push('"' + String(row[col]).replace(/"/g, '""') + '"');
                    } else {
                        cells.push('" "');
                    }
                }
                tbody.push(cells.join(';'));
            }

            table = `${thead}\r\n${tbody.join("\r\n")}`;
        }

        const uri = 'data:application/csv;charset=utf-8;base64,';
        const base64 = (s) => window.btoa(unescape(encodeURIComponent(s)));

        let link = document.createElement("a");
        link.download = "export.csv";
        link.href = uri + base64(table);
        link.click();
    }

    selectAllHandler = (event) => {
        let selected = [];
        if (!this.state.selected.length) {
            for (let i = 0; i < this.state.rows.length; i++) {
                selected.push(i);
            }
        }
        this.setState({ selected: selected })
    }

    defaultSortHandler = (event) => {
        this.setState({
            rows: this.multiSort(this.state.rows, this.defaultSortParams),
            sortParams: this.defaultSortParams,
        });
    }

    render() {
        if (this.state.isEmpty) {
            return <div>{this.props.emptyTableMessage}</div>;
        }

        const {
            cols,
            rows,
            totalPages,
            currentPage
        } = this.state;

        const totalCols = cols.length;
        const totalRows = rows.length;

        const {
            rowsPerPage,
            selectable,
            csvExport
        } = this.props;

        const pageFirstRow = (currentPage - 1) * rowsPerPage;
        const pageLastRow = totalRows < pageFirstRow + rowsPerPage ? totalRows : pageFirstRow + rowsPerPage;

        let cells = [];

        //head
        let thead = [];
        for (const col of cols) {
            let sortObj = this.state.sortParams.find(el => el.name === col.name);
            let sort = '';
            if (sortObj !== undefined) {
                switch (sortObj.dir) {
                    case 'ASC': sort = '\u2197'; break;
                    case 'DESC': sort = '\u2198'; break;
                    default: sort = '';
                }
            }
            cells.push(
                <th
                    key={`th${col.name}`}
                    onMouseDown={event => this.headerMouseDownHandler(event, col.name)}>
                    {[col.title, sort]}
                </th>
            );
        };
        thead = <tr key="thr">{cells}</tr>;

        //footer
        let paginator = null;
        if (totalPages > 1) {
            paginator =
                <Paginator
                    cp={currentPage}
                    tp={totalPages}
                    pageClickHandler={this.pageClickHandler} />;
        }
        let tfoot =
            <tr key="tfr">
                <td colSpan={totalCols}>
                    <div className={classes.Footer}>
                        <div className={classes.Legend}>Rows {pageFirstRow + 1} to {pageLastRow} of {totalRows}</div>
                        {paginator}
                    </div>
                </td>
            </tr>;

        //body
        let tbody = [];
        for (let i = pageFirstRow; i < pageLastRow; i++) {
            let attachedClasses = [];
            cells = [];
            const row = rows[i];

            for (const col of cols) {
                const tdKey = `td${i}${col.name}`;

                let value = col.formatter ? col.formatter(col, row) : row[col.name];

                if (col.html) {
                    cells.push(
                        <td
                            key={tdKey}
                            dangerouslySetInnerHTML={{ __html: value }}>
                        </td>
                    );
                } else {
                    cells.push(
                        <td key={tdKey}>
                            {value}
                        </td>
                    );
                }
            }

            if (this.state.selected.indexOf(i) !== -1) {
                attachedClasses.push(classes.Selected);
            }
            if (i % 2 === 0) {
                attachedClasses.push(classes.Odd);
            }

            tbody.push(
                <tr
                    key={`tbr${i}`}
                    className={attachedClasses.join(' ')}
                    onMouseDown={selectable ? event => this.rowMouseDownHandler(event, i, pageFirstRow, pageLastRow) : null}>
                    {cells}
                </tr>
            );
        }

        return (
            <div
                className={classes.Container}
                onMouseEnter={this.toolbarShow}
                onMouseLeave={this.toolbarHide}>
                <table
                    className={classes.Table}>
                    <thead ref={this.refHeader} onMouseEnter={this.toolbarShow}>{thead}</thead>
                    <tfoot onMouseEnter={this.toolbarHide}>{tfoot}</tfoot>
                    <tbody ref={this.refBody} onMouseEnter={this.toolbarShow}>{tbody}</tbody>
                </table>
                <Toolbar
                    show={this.state.toolbarShow}
                    defaultTop={this.state.headerHeight}
                    boundaryTop={this.state.bodyTop}
                    boundaryBottom={this.state.bodyTop + this.state.bodyHeight}
                    defaultSortHandler={this.defaultSortHandler}
                    csvExportHandler={csvExport ? this.csvExportHandler : null}
                    selectAllHandler={selectable ? this.selectAllHandler : null} />
            </div>
        );
    }
}
