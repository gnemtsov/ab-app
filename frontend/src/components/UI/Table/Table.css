.Table tbody td{
    padding: 0.4rem;
}

.Odd{
    background-color: #e9e8e8;
}

.Selected{
    background-color: aquamarine;
}

.Footer{
    display: flex;
    justify-content: space-between;
    align-items: center;
    flex-wrap: wrap-reverse;
}

.Footer div{
    margin-top: 1rem;
}

.Footer div:first-child {
    margin-right: 2rem;
}