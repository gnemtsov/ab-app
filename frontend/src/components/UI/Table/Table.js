import React, { Component } from 'react';
import Aux from '../../../hoc/Auxillary/Auxillary';
import Paginator from './Paginator/Paginator';

import * as utility from '../../../shared/utility';

import classes from './Table.css';

export class Table extends Component {

    state = {
        data: {
            conf: {
                rowsPerPage: 15,
                selectable: false
            },
            cols: [],
            rows: []
        },
        totalPages: 1,
        currentPage: 1,
        selected: [],
        sortParams: [],
        defaultSortParams: []
    }

    constructor(props) { //constructor updates initial state
        super(props);

        const data = props.data; //table data provided
        if (data === undefined || data.cols === undefined || !data.cols.length) {
            return;
        }

        data.conf = {
            ...this.state.data.conf,
            ...data.conf
        }

        const totalPages = Math.ceil(data.rows.length / data.conf.rowsPerPage);

        let defaultSortParams = [];
        let sortCols = data.cols.filter(col => col.sortDirection !== undefined);
        if (sortCols.length) {
            sortCols.sort((a, b) => a.sortOrder - b.sortOrder);

            defaultSortParams = sortCols.map((col) => ({
                name: col.name,
                dir: col.sortDirection
            }));
        }

        if (defaultSortParams.length) {
            const cols = defaultSortParams.map((col) => col.name);
            const dirs = defaultSortParams.map((col) => col.dir);
            data.rows = utility.multiSort(data.rows, cols, dirs);
        }

        this.state = {
            ...this.state,
            data,
            totalPages,
            sortParams: defaultSortParams,
            defaultSortParams
        };
    }

    exportToCSV = () => { //export data as comma-separated text
        const data = this.state.data;
        
        let table = 'No data specified';
        if (data !== undefined && !data.length) {
            //head
            const thead = data.cols.map(col => '"' + String(col.title) + '"').join(';');

            //body
            let tbody = [];
            for (let i = 0; i < data.rows.length; i++) {
                let cells = [];
                const row = data.rows[i];
                for (var col in row) {
                    if (row.hasOwnProperty(col)) {
                        cells.push('"' + String(row[col]).replace(/"/g, '""') + '"');
                    } else {
                        cells.push('" "');
                    }
                }
                tbody.push(cells.join(';'));
            }

            table = `${thead}\r\n${tbody.join("\r\n")}`;
        }

        const uri = 'data:application/csv;charset=utf-8;base64,';
        const base64 = (s) => window.btoa(unescape(encodeURIComponent(s)));

        let link = document.createElement("a");
        link.download = "export.csv";
        link.href = uri + base64(table);
        link.click();
    }

    headerMouseDownHandler = (event, colName) => { //sort handler
        event.preventDefault();

        let sortParams = this.state.sortParams.slice();
        let sortIndex = sortParams.findIndex(el => el.name === colName);

        if (event.shiftKey) {
            if (sortIndex === -1) {
                sortParams.push({
                    name: colName,
                    dir: 'ASC'
                });
            } else if (sortParams[sortIndex].dir === 'ASC') {
                sortParams[sortIndex].dir = 'DESC';
            } else {
                sortParams.splice(sortIndex, 1);
            }
        } else {
            if (sortIndex === -1) {
                sortParams = [{
                    name: colName,
                    dir: 'ASC'
                }];
            } else if (sortParams[sortIndex].dir === 'ASC') {
                sortParams = [{
                    name: colName,
                    dir: 'DESC'
                }];
            } else {
                sortParams = [];
            }
        }

        const cols = sortParams.map((col) => col.name);
        const dirs = sortParams.map((col) => col.dir);
        const rows = this.state.data.rows.slice();
        const sortedRows = utility.multiSort(rows, cols, dirs);
        const updatedSelected = this.state.selected.map(i => sortedRows.indexOf(rows[i])); //remap selected values after sort

        this.setState({
            ...this.state,
            data: {
                ...this.state.data,
                rows: sortedRows
            },
            selected: updatedSelected,
            sortParams: sortParams
        });
    }

    rowMouseDownHandler = (event, row, pageFirstRow, pageLastRow) => { //select handler

        let selected = this.state.selected.slice();

        if (event.ctrlKey) {
            const selectedRowIndex = selected.indexOf(row);
            selectedRowIndex === -1 ? selected.push(row) : selected.splice(selectedRowIndex, 1);
        } else if (event.shiftKey) {
            event.preventDefault();
            const selectedRowIndex = selected.indexOf(row);
            const selectedOnPage = selected.filter(i => i >= pageFirstRow && i <= pageLastRow);
            const lastSelectedRow = selectedOnPage.length ? selectedOnPage[selectedOnPage.length - 1] : row;

            const selectAction = (i) => {
                const selectedI = selected.indexOf(i);
                if (selectedRowIndex === -1) {
                    selectedI === -1 && selected.push(i);
                } else {
                    row !== i && selected.splice(selectedI, 1);
                }
            }

            if (row > lastSelectedRow) {
                for (let i = lastSelectedRow; i <= row; i++) {
                    selectAction(i);
                }
            } else {
                for (let i = lastSelectedRow; i >= row; i--) {
                    selectAction(i);
                }
            }
        } else {
            selected = selected.filter(row => row < pageFirstRow || row > pageLastRow);
            selected.push(row);
        }

        this.setState({
            ...this.state,
            selected: selected
        });
    }

    pageClickHandler = (event, page) => { //paginator handler
        event.preventDefault();

        let newPage = page;
        if (newPage === 'back') {
            newPage = this.state.currentPage - 1;
        } else if (newPage === 'forward') {
            newPage = this.state.currentPage + 1;
        }

        this.setState({
            ...this.state,
            currentPage: newPage
        });
    }

    render() {
        const data = this.state.data;
        console.log('Components, UI, Table: rendering', data);

        let table = 'No data specified';
        if (data !== undefined && !data.length) {
            let thead = [];
            let tbody = [];

            //head
            let cells = [];
            data.cols.forEach((col, key) => {
                let sortObj = this.state.sortParams.find(el => el.name === col.name);
                let sort = '';
                if (sortObj !== undefined) {
                    switch (sortObj.dir) {
                        case 'ASC': sort = '\u2197'; break;
                        case 'DESC': sort = '\u2198'; break;
                        default: sort = '';
                    }
                }
                cells.push(
                    <th
                        key={`th${key}`}
                        onMouseDown={event => this.headerMouseDownHandler(event, col.name)}>
                        {[col.title, sort]}
                    </th>
                );
            });
            thead = <tr key="thr">{cells}</tr>;

            //body
            const pageFirstRow = (this.state.currentPage - 1) * data.conf.rowsPerPage;
            const pageLastRow = data.rows.length < pageFirstRow + data.conf.rowsPerPage ? data.rows.length : pageFirstRow + data.conf.rowsPerPage;
            for (let i = pageFirstRow; i < pageLastRow; i++) {
                cells = [];
                const row = data.rows[i];
                for (var col in row) {
                    if (row.hasOwnProperty(col)) {
                        cells.push(<td key={`td${i}${col}`}>{row[col]}</td>);
                    } else {
                        cells.push(<td key={`td${i}${col}`}></td>);
                    }
                }
                tbody.push(
                    <tr
                        key={`tbr${i}`}
                        className={this.state.selected.indexOf(i) === -1 ? '' : classes.Selected}
                        onMouseDown={data.conf.selectable ? event => this.rowMouseDownHandler(event, i, pageFirstRow, pageLastRow) : null}>
                        {cells}
                    </tr>
                );
            }

            table = (
                <table>
                    <thead>{thead}</thead>
                    <tbody>{tbody}</tbody>
                </table>
            );
        }

        return (
            <Aux>
                {table}
                <button onClick={this.exportToCSV}>Export to csv</button>
                <Paginator currentPage={this.state.currentPage} totalPages={this.state.totalPages} pageClickHandler={this.pageClickHandler} />
            </Aux>
        );
    }
}

export default Table;