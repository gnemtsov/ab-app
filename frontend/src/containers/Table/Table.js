import React, { Component } from 'react';
import axios from 'axios';
import memoize from 'memoize-one';

import ErrorBoundary from '../../hoc/errorBoundary/errorBoundary';
import Spinner from '../../components/UI/Spinner/Spinner';
import TableComponent from '../../components/Table/Table';

import * as Formatters from './Formatters/Formatters';

export class Table extends Component {

    conf = {
        selectable: true
    }
    
    formattersToFunctions = memoize (
		(cols) => cols.map( col => { //make functions out of formatters
			const newCol = {...col};
			if (newCol.frontendFormatter !== undefined) {
				newCol.formatter = Formatters[newCol.frontendFormatter];
				delete newCol.frontendFormatter;
			}
			return newCol;
		})
    );

    constructor(props) {
        super(props);

        this.conf = {
            ...this.conf,
            ...props.conf
        }
    }

    componentDidMount() {
        /*if (this.state.cols === undefined) {
            axios.get(this.props.api)
                .then(result => this.setState(...result.data));
        }*/
    }

    render() {
        let table = <Spinner />;
		let {cols, rows} = this.props;
        if (cols !== undefined) {
			cols = this.formattersToFunctions(cols);
            table = <TableComponent {...this.conf} cols={cols} rows={rows} />;
        }

        return (
            <React.Fragment>
                <h1>{this.props.title}</h1>
                <ErrorBoundary>{table}</ErrorBoundary>
            </React.Fragment >
        );
    }
}

export default Table;
