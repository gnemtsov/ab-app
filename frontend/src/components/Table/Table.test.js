import React from 'react';

import { configure, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Table from './Table';

configure({ adapter: new Adapter() });

describe('Table', () => {

    const snapshoot = (name, conf, cols, rows) => {
        const wrapper = render(
            <Table
                {...conf}
                cols={cols}
                rows={rows} />
        );
        it(name, () => {
            expect(wrapper).toMatchSnapshot();
        });
    }

    let conf = {};
    let cols = [
        { name: 'name', title: 'Pet name' },
        { name: 'class', title: 'Animal class' },
        { name: 'age', title: 'Age' },
        { name: 'gender', title: 'Gender' }
    ];
    let rows = [
        { name: 'Buddy', class: 'Dog', age: 3, gender: 'male' },
        { name: 'Molly', class: 'Cat', age: 15, gender: 'female' },
        { name: 'Bonnie', class: 'Cat', age: 2, gender: 'female' },
        { name: 'Coco', class: 'Parrot', age: 22, gender: 'male' },
        { name: 'Oscar', class: 'Dog', age: 5, gender: 'male' },
        { name: 'Max', class: 'Turtle', age: 15, gender: 'male' },
        { name: 'Jack', class: 'Varan', age: 1, gender: 'male' }
    ];

    snapshoot('Basic render test', conf, cols, rows);

    cols = [
        { name: 'name', title: 'Pet name' },
        { name: 'class', title: 'Animal class', sortOrder: 1, sortDirection: 'ASC' },
        { name: 'age', title: 'Age', sortOrder: 2, sortDirection: 'ASC' },
        { name: 'gender', title: 'Gender' }
    ];
    snapshoot('With sorting', conf, cols, rows);

    conf = {
        rowsPerPage: 2,
        selectable: true
    };
    snapshoot('With pagination', conf, cols, rows);

    rows = [];
    snapshoot('With no data', conf, cols, rows);


});