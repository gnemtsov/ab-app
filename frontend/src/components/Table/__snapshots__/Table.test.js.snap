// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Basic render test 1`] = `
<div
  class="Container"
>
  <table
    class="Table"
  >
    <thead>
      <tr>
        <th>
          Pet name
        </th>
        <th>
          Animal class
        </th>
        <th>
          Age
        </th>
        <th>
          Gender
        </th>
      </tr>
    </thead>
    <tfoot>
      <tr>
        <td
          colspan="4"
        >
          <div
            class="Footer"
          >
            <div
              class="Legend"
            >
              Rows 1 to 7 of 7
            </div>
          </div>
        </td>
      </tr>
    </tfoot>
    <tbody>
      <tr
        class="Odd"
      >
        <td>
          Buddy
        </td>
        <td>
          Dog
        </td>
        <td>
          3
        </td>
        <td>
          male
        </td>
      </tr>
      <tr
        class=""
      >
        <td>
          Molly
        </td>
        <td>
          Cat
        </td>
        <td>
          15
        </td>
        <td>
          female
        </td>
      </tr>
      <tr
        class="Odd"
      >
        <td>
          Bonnie
        </td>
        <td>
          Cat
        </td>
        <td>
          2
        </td>
        <td>
          female
        </td>
      </tr>
      <tr
        class=""
      >
        <td>
          Coco
        </td>
        <td>
          Parrot
        </td>
        <td>
          22
        </td>
        <td>
          male
        </td>
      </tr>
      <tr
        class="Odd"
      >
        <td>
          Oscar
        </td>
        <td>
          Dog
        </td>
        <td>
          5
        </td>
        <td>
          male
        </td>
      </tr>
      <tr
        class=""
      >
        <td>
          Max
        </td>
        <td>
          Turtle
        </td>
        <td>
          15
        </td>
        <td>
          male
        </td>
      </tr>
      <tr
        class="Odd"
      >
        <td>
          Jack
        </td>
        <td>
          Varan
        </td>
        <td>
          1
        </td>
        <td>
          male
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="Container"
  >
    <div
      class="Toolbar"
      style="top:0px"
    >
      <button
        class="Button"
      >
        ⇅
      </button>
      <button
        class="Button"
      >
        csv
      </button>
    </div>
  </div>
</div>
`;

exports[`Table With no data 1`] = `
<div>
  No data specified
</div>
`;

exports[`Table With pagination 1`] = `
<div
  class="Container"
>
  <table
    class="Table"
  >
    <thead>
      <tr>
        <th>
          Pet name
        </th>
        <th>
          Animal class↗
        </th>
        <th>
          Age↗
        </th>
        <th>
          Gender
        </th>
      </tr>
    </thead>
    <tfoot>
      <tr>
        <td
          colspan="4"
        >
          <div
            class="Footer"
          >
            <div
              class="Legend"
            >
              Rows 1 to 2 of 7
            </div>
            <div
              class="Paginator"
            >
              <button
                class="Button"
                disabled=""
              >
                ←
              </button>
              <button
                class="Button"
                disabled=""
              >
                1
              </button>
              <button
                class="Button"
              >
                2
              </button>
              <button
                class="Button"
              >
                3
              </button>
              <button
                class="Button"
              >
                4
              </button>
              <button
                class="Button"
              >
                →
              </button>
            </div>
          </div>
        </td>
      </tr>
    </tfoot>
    <tbody>
      <tr
        class="Odd"
      >
        <td>
          Bonnie
        </td>
        <td>
          Cat
        </td>
        <td>
          2
        </td>
        <td>
          female
        </td>
      </tr>
      <tr
        class=""
      >
        <td>
          Molly
        </td>
        <td>
          Cat
        </td>
        <td>
          15
        </td>
        <td>
          female
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="Container"
  >
    <div
      class="Toolbar"
      style="top:0px"
    >
      <button
        class="Button"
      >
        ⇅
      </button>
      <button
        class="Button"
      >
        csv
      </button>
      <button
        class="Button"
      >
        ✽
      </button>
    </div>
  </div>
</div>
`;

exports[`Table With sorting 1`] = `
<div
  class="Container"
>
  <table
    class="Table"
  >
    <thead>
      <tr>
        <th>
          Pet name
        </th>
        <th>
          Animal class↗
        </th>
        <th>
          Age↗
        </th>
        <th>
          Gender
        </th>
      </tr>
    </thead>
    <tfoot>
      <tr>
        <td
          colspan="4"
        >
          <div
            class="Footer"
          >
            <div
              class="Legend"
            >
              Rows 1 to 7 of 7
            </div>
          </div>
        </td>
      </tr>
    </tfoot>
    <tbody>
      <tr
        class="Odd"
      >
        <td>
          Bonnie
        </td>
        <td>
          Cat
        </td>
        <td>
          2
        </td>
        <td>
          female
        </td>
      </tr>
      <tr
        class=""
      >
        <td>
          Molly
        </td>
        <td>
          Cat
        </td>
        <td>
          15
        </td>
        <td>
          female
        </td>
      </tr>
      <tr
        class="Odd"
      >
        <td>
          Buddy
        </td>
        <td>
          Dog
        </td>
        <td>
          3
        </td>
        <td>
          male
        </td>
      </tr>
      <tr
        class=""
      >
        <td>
          Oscar
        </td>
        <td>
          Dog
        </td>
        <td>
          5
        </td>
        <td>
          male
        </td>
      </tr>
      <tr
        class="Odd"
      >
        <td>
          Coco
        </td>
        <td>
          Parrot
        </td>
        <td>
          22
        </td>
        <td>
          male
        </td>
      </tr>
      <tr
        class=""
      >
        <td>
          Max
        </td>
        <td>
          Turtle
        </td>
        <td>
          15
        </td>
        <td>
          male
        </td>
      </tr>
      <tr
        class="Odd"
      >
        <td>
          Jack
        </td>
        <td>
          Varan
        </td>
        <td>
          1
        </td>
        <td>
          male
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="Container"
  >
    <div
      class="Toolbar"
      style="top:0px"
    >
      <button
        class="Button"
      >
        ⇅
      </button>
      <button
        class="Button"
      >
        csv
      </button>
    </div>
  </div>
</div>
`;
